LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -30,96 +27,114 @@ internal sealed class ConcentratorDeduplication<T> :
         {
             this.cache = cache;
-            this.socketRegistry = socketRegistry;
             this.logger = logger;
         }
 
-        public bool ShouldDrop(T frame, StationEui stationEui)
+        public ConcentratorDeduplicationResult CheckDuplicateJoin(LoRaRequest loRaRequest)
         {
-            if (frame == null) throw new ArgumentNullException(nameof(frame));
+            _ = loRaRequest ?? throw new ArgumentNullException(nameof(loRaRequest));
+            var key = CreateCacheKey((LoRaPayloadJoinRequest)loRaRequest.Payload);
+            if (EnsureFirstMessageInCache(key, loRaRequest, out _))
+                return ConcentratorDeduplicationResult.NotDuplicate;
+
+            this.logger.LogDebug($"{Constants.DuplicateMessageFromAnotherStationMsg} with EUI {loRaRequest.StationEui}.");
+            return ConcentratorDeduplicationResult.Duplicate;
+        }
 
-            var key = CreateCacheKey(frame);
+        public ConcentratorDeduplicationResult CheckDuplicateData(LoRaRequest loRaRequest, LoRaDevice loRaDevice)
+        {
+            _ = loRaRequest ?? throw new ArgumentNullException(nameof(loRaRequest));
+            _ = loRaDevice ?? throw new ArgumentNullException(nameof(loRaDevice));
+            var key = CreateCacheKey((LoRaPayloadData)loRaRequest.Payload);
+            if (EnsureFirstMessageInCache(key, loRaRequest, out var previousStation))
+                return ConcentratorDeduplicationResult.NotDuplicate;
 
-            StationEui previousStation;
-            lock (this.cache)
+            if (previousStation == loRaRequest.StationEui)
             {
-                if (!this.cache.TryGetValue(key, out previousStation))
-                {
-                    AddToCache(key, stationEui);
-                    return false;
-                }
+                this.logger.LogDebug($"Message was received previously from the same EUI {loRaRequest.StationEui} (\"confirmedResubmit\").");
+                return ConcentratorDeduplicationResult.DuplicateDueToResubmission;
             }
 
-            if (previousStation == stationEui)
+            if (loRaDevice.Deduplication == DeduplicationMode.Drop)
             {
-                // considered as a resubmit
-                this.logger.LogDebug("Message received from the same EUI {StationEui} as before, will not drop.", stationEui);
-                return false;
+                this.logger.LogDebug($"{Constants.DuplicateMessageFromAnotherStationMsg} with EUI {loRaRequest.StationEui}.");
+                return ConcentratorDeduplicationResult.Duplicate;
             }
 
-            // received from a different station
-            if (this.socketRegistry.IsSocketWriterOpen(previousStation))
+            this.logger.LogDebug($"Message from station with EUI {loRaRequest.StationEui} marked as soft duplicate due to DeduplicationStrategy.");
+            return ConcentratorDeduplicationResult.SoftDuplicateDueToDeduplicationStrategy;
```

Comment:
Let's simplify this:

```
            _ = loRaRequest ?? throw new ArgumentNullException(nameof(loRaRequest));
            _ = loRaDevice ?? throw new ArgumentNullException(nameof(loRaDevice));
            var key = CreateCacheKey((LoRaPayloadData)loRaRequest.Payload);
            if (EnsureFirstMessageInCache(key, loRaRequest, out var previousStation))
                return ConcentratorDeduplicationResult.NotDuplicate;

            var result = ConcentratorDeduplicationResult.SoftDuplicateDueToDeduplicationStrategy;

            if (previousStation == loRaRequest.StationEui)
            {
                result = ConcentratorDeduplicationResult.DuplicateDueToResubmission;
            }
            else if (loRaDevice.Deduplication == DeduplicationMode.Drop)
            {
                result = ConcentratorDeduplicationResult.Duplicate;
            }

            this.logger.LogDebug($"{Constants.DuplicateMessageFromAnotherStationMsg} {result}. Message was sent from station {loRaRequest.StationEui}.");
            return result;
```

This way we have only one log statement we need to write, we have a consistent log message for all cases and only the result is different and we can use the constant message to also check on the specific result in the E2E test.

Reply:
0d923f9655c55d7aa762c6a6ba85fc717c2813a5